[WIP] Add new pool join-check: pool and host must have same "enforced" updates

lindig opened the pull request:
This commit adds an additional check before a host can join a pool.
    The joining host compares its set of updates with the updates on each
    host in the pool.  These sets must be equal for the host being able to join
    the pool.  The comparison only considers updates that are marked as
    enforce-homogeneity="true" in their update.xml definition -- see below.
    This attribute is new. The absence of the enforce-homogeneity attribute
    in an update defaults to enforce-homogeneity=false. This makes the code
    backward compatible with updates that don't have this annotation.
    
    Typically, code updates will have enforce-homogeneity=true but driver
    updates will have enforce-homogeneity=false such that hosts in the pool
    can have different sets of drivers but have the same code updates.
    
    The implementation changes the datamodel: class pool_update gets a new
    field enforce_homogeneity.
    
    <update after-apply-guidance="restartHost" build-number="XS71ECU1/58"
      control="control-XS71ECU1" enforce-homogeneity="true"
      installation-size="566138764" key="RPM-GPG-KEY-XS-Eng-Test.pub"
      name-label="XS71ECU1" uuid="5eea06f3-4990-497d-af7a-d0bba6f9f96a"
      version="1.0">
    
      <name-description>Cumulative Update 1 (CU-1)</name-description>
      <rollsup name-label="XS71E001" uuid="fc438a32-0214-4193-8676-9feb121c6997"/>
      <rollsup name-label="XS71E002" uuid="9768c0f1-d111-4ee4-ac85-fe9ee37b5726"/>
      <rollsup name-label="XS71E003" uuid="b80ea320-f9ae-414c-a14d-781861b5c0ac"/>
      <rollsup name-label="XS71E004" uuid="a4ebb0e4-bfcb-48da-8b42-4673d3111d29"/>
      <rollsup name-label="XS71E005" uuid="d2df0c4f-eaf6-4778-a754-19a8b7739b5c"/>
      <rollsup name-label="XS71E006" uuid="008f1b43-8f02-40a4-8475-28f15dbfd1fc"/>
      <rollsup name-label="XS71E007" uuid="04bcb1f7-3bdc-440d-9cac-9d2a87746b24"/>
      <rollsup name-label="XS71E008" uuid="e3c382d7-dba5-4809-9901-d0d15ee02ad9"/>
    </update>
    
    modified:   ocaml/idl/datamodel.ml
    modified:   ocaml/xapi/test_common.ml
    modified:   ocaml/xapi/xapi_pool.ml
    modified:   ocaml/xapi/xapi_pool_update.ml
    
    Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

mseri commented on ocaml/xapi/xapi_pool.ml:
```diff
@@ -143,6 +143,41 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
+      |> List.map (fun upd -> upd.API.pool_update_uuid)
+      |> S.of_list in
+    let updates_local = (* patches on local host *)
+      Db.Host.get_updates ~__context ~self:localhost
```
We start getting this kind of duplication all over the place, would be nice to be able to pass Db/Client with a common api as function parameters...

mseri replied:
However I think you can take out the `filter/map/set_of` into a common function to reduce some of the duplication

mseri commented on ocaml/xapi/xapi_pool.ml:
```diff
@@ -143,6 +143,41 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
+        let updates_remote = updates_of pool_host in
+        let host = Client.Host.get_record rpc session_id pool_host in
+        if not (S.equal updates_local updates_remote) then begin
+          let only_remote = S.diff updates_remote updates_local in
```
It's not used anywhere else, why not:
```ocaml
let get_diff main cmp = S.diff main cmp |> S.elements  |> String.concat "," in
let only_remote = get_diff updates_remote updates_local in
let only_local = get_diff updates_local updates_remote in
let msg  = Printf.sprintf "Updates differ: Only on pool host %s: {%s} -- only on localhost: {%s}" host.API.host_uuid only_remote only_local in
raise Api_errors.(Server_error(pool_hosts_not_homogeneous,[msg]))
```

mseri commented on ocaml/xapi/xapi_pool_update.ml:
```diff
@@ -259,6 +260,13 @@ let parse_update_info xml =
+    let enforce_homogeneity =
+      try
+        List.assoc "enforce-homogeneity" attr
+        |> fun s -> s = "1" || String.lowercase s = "true"
```
I think to remember an helper somewhere in xapi that does exactly this. It would be better to use the helper to reduce the changes when updating compiler: `String.lowercase` has been deprecated and renamed `String.lowercase_ascii` in OCaml 4.03.0

lindig replied:
* https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_vm_lifecycle.ml#L27
* outside xapi: https://github.com/xapi-project/xenopsd/blob/master/lib/platform.ml#L25

mseri replied:
That's the one. Probably we should move it to helpers and use it where needed

lindig replied:
We are currently not on 4.03.0 (I believe) and this can't use `String.lowercase_ascii`.

mseri replied:
I know. I am just saying that having only one copy of this function to call will make it easier to fix the deprecation in the future (and also I like to reduce duplication)

mseri commented on ocaml/xapi/xapi_pool_update.ml:
```diff
@@ -275,6 +283,7 @@ let parse_update_info xml =
       key = Filename.basename key;
       installation_size = installation_size;
       after_apply_guidance = guidance;
+      enforce_homogeneity = enforce_homogeneity;
```
I think we should modernise this code and instead write:
```ocaml
update_info = {
       uuid;
       name_label;
       name_description;
       version;
       key = Filename.basename key;
       installation_size;
       after_apply_guidance = guidance;
       enforce_homogeneity;
}
```

lindig replied:
That gives me permission to do this, right? ;-)
```
update_info = 
  { uuid
  ; name_label
  ; ...
  }

gaborigloi replied:
That's my favourite style too :smile: (learn it from elm-lang: http://elm-lang.org/docs/style-guide )

mseri replied:
Go with it!! :-)

mseri commented on ocaml/xapi/xapi_pool_update.ml:
```diff
@@ -510,7 +520,7 @@ let pool_update_download_handler (req: Request.t) s _ =
   let filepath = String.sub_to_end req.Request.uri (String.length Constants.get_pool_update_download_uri)
                  |> Filename.concat Xapi_globs.host_update_dir
-                 |> Stdext.Unixext.resolve_dot_and_dotdot 
+                 |> Stdext.Unixext.resolve_dot_and_dotdot
```
Each time I see this function name I get a smile :D

lindig commented:
I've addressed the comments:

* A new function `is_true` in `helpers.ml` to to recognize booolean strings - taken from xenopsd.
* Reduced redundancies by using a local rpc call - please review this @jonludlam 
* Minor stylistic cleanup
* Added logging
* The error message now return the UUIDs of the hosts that differ. That seems to me the most
plausible error message. Jon had advised that these messages should not contain string messages that would require i18n.

mseri commented:
I am happy with the changes. I am not going to approve this because I think that the other people involved with the update should have a final word on it.

thomassa commented on ocaml/xapi/helpers.ml:
```diff
@@ -1312,3 +1312,16 @@ let retry ~__context ~doc ?(policy = Policy.standard) f =
 let retry_with_global_lock ~__context ~doc ?policy f =
   retry ~__context ~doc ?policy (fun () -> with_global_lock f)
+
+  (** [is_true] returns [true] if [key] in string map [assoc] denotes
```
This `is_true` function is an exact duplicate of the function of the same name in `ocaml/xapi/vm_platform.ml` so please can we have just one of them?

Also we have a different `is_true` in `ocaml/xapi/xapi_db_upgrade.ml`
and our own versions of `bool_of_string` in `ocaml/xapi/cli_operations.ml` and `ocaml/xapi/system_domains.ml`
but we might want to keep those for their specific behaviours.

lindig replied:
I've fixed that.

lindig commented:
Quite some updates, please review them:

* This has been tested manually by joining and leaving a pool and adding updates to different sides 
* Local and remote API functions need different rpc and session values - this is now fixed
* The new  enforce-homogeneity flag is now listed in `xe update-list` but not by default
* The error only expects one string - now reporting the UUID of the pool host that is different than the local host
* Improved log message 
* In general the user should not see the error because XC will prevent hosts from joining that are incompatible

jonludlam commented on ocaml/xapi/xapi_pool.ml:
```diff
@@ -143,6 +143,36 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
+        let remote_uuid  = Client.Host.get_uuid rpc session_id pool_host in
+        let diff xs ys   = S.diff xs ys |> S.elements |> String.concat "," in
+        let reason       = [remote_uuid] in
+        info
```
I think this should be error rather than info.

lindig replied:
I would argue that a join failing is a normal outcome of an API call and not an error. However, we would like to know the details - hence `info`.

jonludlam commented:
Lovely. Thanks @lindig
